### PR TITLE
Additional fixes for PMD on Maven.

### PR DESCRIPTION
**Pull request by HSAR**

### CODE REVIEW

File: Tasks/Maven/mavencodeanalysis.ts [hunk 1]
```diff
@@ -8,8 +8,8 @@ import trm = require('vsts-task-lib/toolrunner');
 
 // Lowercased names are to lessen the likelihood of xplat issues
 import pmd = require('./mavenpmd');
-import ar = require('./analysisresult');
-import ma = require('./moduleanalysis');
+import {AnalysisResult} from './analysisresult';
+import {ModuleAnalysis} from './moduleanalysis';
```
Comment (HSAR) [inline Tasks/Maven/mavencodeanalysis.ts]: This change is a Typescript-specific syntax which looks much nicer - all class references below have changed to be more brief and readable.

File: Tasks/Maven/mavencodeanalysis.ts [hunk 2]
```diff
@@ -142,16 +142,11 @@ function processAndAssignAnalysisResults(enabledCodeAnalysisTools:string[], modu
...
+function createAndUploadBuildSummary(enabledTools:string[], modules:ModuleAnalysis[]):void {
     var buildSummaryLines:string[] = [];
-    var totalViolationsInBuild = 0;
 
     enabledTools.forEach((toolName:string) => {
-        var toolAnalysisResults:ar.AnalysisResult[] = getToolAnalysisResults(modules, toolName);
-
-        toolAnalysisResults.forEach((toolAnalysisResult:ar.AnalysisResult) => {
-            totalViolationsInBuild += toolAnalysisResult.totalViolations;
-        });
```
Comment (HSAR) [inline Tasks/Maven/mavencodeanalysis.ts]: Non-structural refactor - moved the calculation of total violations out of this method to make it more readable.

File: Tasks/Maven/mavencodeanalysis.ts [hunk 3]
```diff
@@ -163,14 +158,16 @@ function createAndUploadBuildSummary(enabledTools:string[], modules:ma.ModuleAna
...
         buildSummaryLines.push("");
         buildSummaryLines.push("Code analysis results can be found in the 'Artifacts' tab.");
     }
 
     // Save and upload build summary
     // Looks like: "PMD found 13 violations in 4 files.  \r\n
-    // FindBugs found 10 violations in 8 files."
+    // FindBugs found 10 violations in 8 files.  \r\n
+    //   \r\n
+    // Code analysis results can be found in the 'Artifacts' tab.  \r\n"
```
Comment (HSAR) [inline Tasks/Maven/mavencodeanalysis.ts]: Updated this comment.

File: Tasks/Maven/mavencodeanalysis.ts [hunk 4]
```diff
@@ -196,11 +193,21 @@ function getToolAnalysisResults(modules:ma.ModuleAnalysis[], toolName:string):ar
...
 }
 
+function getTotalViolationsInModules(modules:ModuleAnalysis[]):number {
+    var totalViolationsInBuild:number = 0;
+    modules.forEach((module:ModuleAnalysis) => {
+        for (var toolName in module.analysisResults) { // The for-in loop gives the keys, not the values
+            totalViolationsInBuild += module.analysisResults[toolName].totalViolations;
+        }
+    });
+    return totalViolationsInBuild;
```
Comment (HSAR) [inline Tasks/Maven/mavencodeanalysis.ts]: The destination of the refactor referred to earlier.

File: Tasks/Maven/mavencodeanalysis.ts [hunk 5]
```diff
@@ -231,36 +238,36 @@ function createSummaryLine(toolName:string, analysisResults:ar.AnalysisResult[])
...
+function uploadBuildArtifactsFromModule(toolName:string, moduleAnalysis:ModuleAnalysis):void {
+    var analysisResult:AnalysisResult;
     if (moduleAnalysis.analysisResults[toolName]) {
-        filesToUpload = moduleAnalysis.analysisResults[toolName].filesToUpload;
+        analysisResult = moduleAnalysis.analysisResults[toolName];
     }
 
-    if (filesToUpload.length < 1) {
+    // If there are no files to upload or there were no violations, return early
+    if (!analysisResult.filesToUpload || analysisResult.filesToUpload.length < 1 || analysisResult.totalViolations < 1) {
```
Comment (HSAR) [inline Tasks/Maven/mavencodeanalysis.ts]: The actual fix - now that we refer to more than one thing from the module's AnalysisResult, we take an instance of that instead.

Then we check that fileToUpload is not null (mostly for robustness), that it contains elements, and that there are violations in the current AnalysisResult. If there are no violations, there isn't any real reason to upload the artifacts.

File: Tasks/Maven/moduleanalysis.ts [hunk 6]
```diff
@@ -1,7 +1,13 @@
...
 export class ModuleAnalysis {
     moduleName: string;
     rootDirectory: string;
-    analysisResults:any = {}; // A dictionary of toolName:string -> analysisResult:AnalysisResult
+    analysisResults:ToolAnalysisResultsDict = {};
+}
+
+// A dictionary of toolName:string -> analysisResult:AnalysisResult
+interface ToolAnalysisResultsDict {
+    [toolName:string]:AnalysisResult
```
Comment (HSAR) [inline Tasks/Maven/moduleanalysis.ts]: Additional robustness -define the types of the dictionary to be string -> AnalysisResult.

File: Tasks/Maven/mavencodeanalysis.ts [hunk 7]
```diff
@@ -163,14 +158,16 @@ function createAndUploadBuildSummary(enabledTools:string[], modules:ma.ModuleAna
...
+    if (getTotalViolationsInModules(modules) > 0) {
         buildSummaryLines.push("");
         buildSummaryLines.push("Code analysis results can be found in the 'Artifacts' tab.");
     }
 
     // Save and upload build summary
     // Looks like: "PMD found 13 violations in 4 files.  \r\n
-    // FindBugs found 10 violations in 8 files."
+    // FindBugs found 10 violations in 8 files.  \r\n
+    //   \r\n
```
Comment (bgavrilMS) [inline Tasks/Maven/mavencodeanalysis.ts]: What's with all the "\r\n" s ?

Comment (HSAR) [inline Tasks/Maven/mavencodeanalysis.ts]: Line returns. There's a specific combination of two spaces and then a line return (a markdown newline) which I have to specify this way - a multiline string tends to cut one of the two out.

**Comment by bgavrilMS** [comment]
LGTM

**Pull request merged**